src/MinimiseProgramTransformer.cpp

```diff
@@ -210,8 +211,10 @@ bool isValidPermutation(
     bool validMapping = true;
     for (size_t i = 0; i < leftAtoms.size() && validMapping; i++) {
         // match arguments
-        std::vector<AstArgument*> leftArgs = leftAtoms[i]->getAtom()->getArguments();
-        std::vector<AstArgument*> rightArgs = rightAtoms[i]->getAtom()->getArguments();
+        std::vector<AstArgument*> leftArgs =
+                dynamic_cast<AstAtomLiteral*>(leftAtoms[i])->getAtom()->getArguments();
```

Comment:
Since left/rightAtoms only contains AstAtomLiterals it seems that it would be much better if the type was std::vector<AstAtomLiteral*> instead of std::vector<AstLiteral*>. Type checks would happen on insertion, rather than extraction.